File: kernel/platform/src/arch/aarch64/emergency_console.c
#include "emergency_console.h"

#include "boot.h"
#include <stdint.h>

#include <stddef.h>
#include "interrupts.h"
#include "virtual_address_allocator.h"
#include "virtual_memory_manager.h"

// Private register offsets for the PL011
#define UART_DR    (0x00 / 4)
#define UART_FR    (0x18 / 4)
#define UART_IBRD  (0x24 / 4)
#define UART_FBRD  (0x28 / 4)
#define UART_LCRH  (0x2C / 4)
#define UART_CR    (0x30 / 4)
#define UART_IMSC  (0x38 / 4) // Interrupt Mask Set/Clear
#define UART_MIS   (0x40 / 4) // Masked Interrupt Status
#define UART_ICR   (0x44 / 4) // Interrupt Clear Register

// Register bits
#define FR_RXFE    (1 << 4) // Receive FIFO Empty
#define FR_TXFF    (1 << 5) // Transmit FIFO Full
#define INT_RX     (1 << 4) // Receive Interrupt bit

static volatile uint32_t *uart_base = NULL;

#define UART_PHYS_ADDR 0x9000000

#define INIT_FAILED_UART_BASE 0x1234

void emergency_console_init(const uintptr_t serial_base) {
    __asm__ volatile("msr daifset, #2"); // Hard interrupts disable

    if ((uintptr_t) uart_base == INIT_FAILED_UART_BASE) {
        // Double-faulting... already tried to initialize the emergency console...
        hcf();
    }
    uart_base = (void *) INIT_FAILED_UART_BASE;
    const uintptr_t virtual_base = vaa_alloc_range(VMM_PAGE_SIZE);

    if (!vmm_map_page(
        &g_kernel_context,
        virtual_base,
        serial_base,
        VMM_FLAG_PRESENT | VMM_FLAG_WRITE | VMM_FLAG_DEVICE
    )) {
        hcf();
    }
    uart_base = (volatile uint32_t *) virtual_base;

    // Hardware Reset - Disable everything first
    uart_base[UART_CR] = 0;
    uart_base[UART_ICR] = 0x7FF; // Clear all sticky interrupts

    // 8N1 + Enable FIFOs. We enable FIFOs here so hardware starts
    // buffering keys even before we enable interrupts.
    uart_base[UART_LCRH] = (3 << 5) | (1 << 4);

    // Enable UART, TX, and RX (Polling mode is now active)
    uart_base[UART_CR] = (1 << 0) | (1 << 8) | (1 << 9);

    emergency_console_println("=========================");
    emergency_console_println("    Emergency Console    ");
    emergency_console_println("=========================");
    emergency_console_println("");
}

static int is_transmit_empty() {
    if (!uart_base)
        return 0;
    return !(uart_base[UART_FR] & FR_TXFF);
}

static void write_serial(char a) {
    if (uart_base == NULL) {
        emergency_console_init(UART_PHYS_ADDR);
    }

    if (a == '\n') {
        write_serial('\r');
    }

    while (!is_transmit_empty()) {
        __asm__ volatile("yield");
    }

    uart_base[UART_DR] = (uint32_t) a;
}

void emergency_console_write(const uint8_t byte) {
    if (!uart_base) {
        emergency_console_init(UART_PHYS_ADDR);
    }

    if (byte == '\n') {
        write_serial('\r');
    }

    while (!is_transmit_empty()) {
        __asm__ volatile("yield");
    }

    uart_base[UART_DR] = (uint32_t) byte;
}

void emergency_console_print(const char *message) {
    if (!message)
        return;
    for (size_t i = 0; message[i] != '\0'; i++) {
        write_serial(message[i]);
    }
}

void emergency_console_println(const char *message) {
    emergency_console_print(message);
    write_serial('\n');
}

void emergency_console_print_hex_u64(const uint64_t value) {
    static const char *hex = "0123456789abcdef";
    emergency_console_print("0x");
    for (int i = 60; i >= 0; i -= 4) {
        uint8_t nib = (value >> i) & 0xF;
        write_serial(hex[nib]);
    }
}

File: kernel/old_kernel/hal/src/arch/aarch64/emergency_console.rs
use crate::arch::cpu::hcf;
use crate::mmu;
use core::arch::asm;
use mmu::VirtualMemoryMappingFlags;

const UART_DR: usize = 0x00;
const UART_FR: usize = 0x06;
const UART_IBRD: usize = 0x09;
const UART_FBRD: usize = 0x10;
const UART_LCRH: usize = 0x2C / 4;
const UART_CR: usize = 0x30 / 4;
const UART_IMSC: usize = 0x38 / 4; // Interrupt Mask Set/Clear
const UART_MIS: usize = 0x40 / 4; // Masked Interrupt Status
const UART_ICR: usize = 0x44 / 4; // Interrupt Clear Register

// Register bits
const FR_RXFE: u32 = 1 << 4; // Receive FIFO Empty
const FR_TXFF: u32 = 1 << 5; // Transmit FIFO Full
const FR_BUSY: u32 = 1 << 3;
const INT_RX: u32 = 1 << 4; // Receive Interrupt bit

pub unsafe fn init(serial_base: u64) -> u64 {
    unsafe {
        asm!("msr daifset, #3", "dmb sy");

        let virtual_base =
            kernel_bindings_gen::vaa_alloc_range(kernel_bindings_gen::VMM_PAGE_SIZE as usize);

        let kernel_context = mmu::get_kernel_context();
        if !mmu::map_page(
            kernel_context,
            virtual_base,
            serial_base as usize,
            VirtualMemoryMappingFlags::PRESENT
                | VirtualMemoryMappingFlags::WRITE
                | VirtualMemoryMappingFlags::DEVICE,
        ) {
            hcf();
        }

        let uart_base = virtual_base as *mut u32;

        // Hardware Reset - Disable everything first
        uart_base.add(UART_CR).write_volatile(0);
        uart_base.add(UART_ICR).write_volatile(0x7FF); // Clear all sticky interrupts

        // 8N1 + Enable FIFOs. We enable FIFOs here so hardware starts
        // buffering keys even before we enable interrupts.
        uart_base.add(UART_LCRH).write_volatile((3 << 5) | (1 << 4));

        // Enable UART, TX, and RX (Polling mode is now active)
        uart_base
            .add(UART_CR)
            .write_volatile((1 << 0) | (1 << 8) | (1 << 9));

        uart_base as u64
    }
}

pub unsafe fn disable(serial_base: u64) {
    unsafe {
        let uart_base = serial_base as *mut u32;

        // Wait for the UART to no longer be busy
        // This ensures the FIFO and the transmit shifter are both empty.
        while (uart_base.add(UART_FR).read_volatile() & FR_BUSY) != 0 {
            asm!("yield");
        }

        // Disable UART, TX, and RX in the Control Register
        uart_base.add(UART_CR).write_volatile(0);

        // Mask all interrupts just in case
        uart_base.add(UART_IMSC).write_volatile(0);

        // Clear any remaining sticky interrupt flags
        uart_base.add(UART_ICR).write_volatile(0x7FF);
    }
}

unsafe fn is_transmit_empty(uart_base: *mut u32) -> bool {
    (uart_base.add(UART_FR).read_volatile() & FR_TXFF) == 0
}

pub unsafe fn write(serial_base: u64, byte: u8) {
    unsafe {
        let uart_base = serial_base as *mut u32;

        if byte == ('\n' as u8) {
            write(serial_base, '\r' as u8);
        }

        while !is_transmit_empty(uart_base) {
            asm!("yield");
        }

        uart_base.add(UART_DR).write_volatile(byte as u32);
    }
}
